compiler/hie-core/src/Development/IDE/Core/Service.hs

```diff
@@ -68,32 +70,24 @@ setProfiling opts shakeOpts =
 shutdown :: IdeState -> IO ()
 shutdown = shakeShut
 
--- | Run a single action using the supplied service. See `runActions`
--- for more details.
-runAction :: IdeState -> Action a -> IO a
-runAction service action = head <$> runActions service [action]
-
--- | Run a list of actions in parallel using the supplied service.
--- This will return as soon as the results of the actions are
+-- This will return as soon as the result of the action is
 -- available.  There might still be other rules running at this point,
 -- e.g., the ofInterestRule.
-runActions :: IdeState -> [Action a] -> IO [a]
-runActions x acts = do
-    var <- newBarrier
-    _ <- shakeRun x acts (signalBarrier var)
-    waitBarrier var
-
--- | This is a synchronous variant of `runAction`. See
--- `runActionsSync` of more details.
-runActionSync :: IdeState -> Action a -> IO a
-runActionSync s a = head <$> runActionsSync s [a]
+runAction :: IdeState -> Action a -> IO a
+runAction ide action = do
+    bar <- newBarrier
+    res <- shakeRun ide [do v <- action; liftIO $ signalBarrier bar v; return v]
+    -- shakeRun might throw an exception, in which case res will finish first
```

Comment:
This comment seems slightly off, if `shakeRun` throws an exception `runAction` will terminate with an exception. Presumably you mean that the action passed to `shakeRun` throws an exception?

Reply:
Hard to argue with more comments here. shakeRun also gets actions that are default actions, like ofInterest, and if they throw exceptions it could take out `action`.